filter: Add -H/--hide option to hide functions without subtree

honggyukim opened the pull request:
filter: Add -H/--hide option to hide functions without subtree

The -H/--hide option is used to apply notrace filter to some functions without
affecting their subtree functions.  In other words, it simply hides
single depth of function and continue to trace child functions.

This option works only for replay time and 'hide' can be used as a
trigger just like other triggers such as filter.
```
  $ gcc -pg -o t-abc tests/s-abc.c
  $ uftrace -H main -T c@hide t-abc
  # DURATION     TID     FUNCTION
              [ 13037] | a() {
              [ 13037] |   b() {
     2.600 us [ 13037] |     getpid();
     8.500 us [ 13037] |   } /* b */
    11.400 us [ 13037] | } /* a */
```
This can be useful when user functions are called from a library, but
don't want to trace functions in the library as the following example.
```
  $ cat call_once.cc
  #include <mutex>
  std::once_flag flag;
  void foo() {}
  int main() {
      std::call_once(flag, foo);
  }

  $ g++ -pg -O2 -pthread call_once.cc
  $ uftrace record a.out
  $ uftrace replay
  # DURATION     TID     FUNCTION
              [ 19661] | main() {
              [ 19661] |   pthread_once() {
              [ 19661] |     std::call_once::$_1::_FUN() {
     1.900 us [ 19661] |       foo();
     6.700 us [ 19661] |     } /* std::call_once::$_1::_FUN */
    50.900 us [ 19661] |   } /* pthread_once */
    56.100 us [ 19661] | } /* main */
```
We can simply hide functions in C++ std namespace easily as follows.
```
  $ uftrace replay -H ^std:: a.out
  # DURATION     TID     FUNCTION
              [ 19666] | main() {
              [ 19666] |   pthread_once() {
     2.300 us [ 19666] |     foo();
    31.400 us [ 19666] |   } /* pthread_once */
    38.000 us [ 19666] | } /* main */
```
It also supports 'hide' trigger for the same purpose.

Closes: #1169

Signed-off-by: Honggyu Kim <honggyu.kp@gmail.com>
Signed-off-by: Namhyung Kim <namhyung@gmail.com>

honggyukim commented:
@namhyung I still have to add a test and also need to update the document, but I see that the name **unpatch** is a bit weird so didn't add it yet. Could you tell me how you think about it?

namhyung commented:
Well I don't know.. but basically it **reuses** the option and there's nothing different if we hide the option from the documentation.

honggyukim commented:
Okay. I will update  the document using the existing name.

honggyukim commented:
I've updated the patch with a test and document. Please have a look. Thanks.

namhyung commented:
Sorry for late reply.  I've checked the code and I think we should implement it only for the replay part now.  I plan to apply it for dynamic tracing later..

namhyung commented:
Also we might consider reusing -P option too.  If it's used we can think all other functions should be hidden (this is the same way we use it for dynamic tracing).

honggyukim commented:
> I think we should implement it only for the replay part now. I plan to apply it for dynamic tracing later..

Sure. I've removed record part from the patch and updated it again.

> Also we might consider reusing -P option too. If it's used we can think all other functions should be hidden (this is the same way we use it for dynamic tracing). But this can be done later..

Right. I will think it may be done by setting `-F FUNC@depth=1`, but need to think about it more. Thanks.

namhyung commented:
Thanks for the update... but I'm still hesitating about the naming, sorry.

Maybe we can add 'hide' as an alias for the trigger and keep the internal naming like it (ex. TRIGGER_FL_HIDE).  It's essentially for hide but we are reusing the -U option.  And then 'unpatch' trigger should work too.

namhyung commented on doc/uftrace-replay.md:
```diff
@@ -70,6 +70,11 @@ COMMON OPTIONS
+:   Set filter not to trace selected functions.  It doesn't affects their
+    subtrees, but hides only the given functions.
+    This option can be used more than once.
+
```
I think we should update the trigger part too.

namhyung commented on uftrace.c:
```diff
@@ -198,7 +198,7 @@ __used static const char uftrace_help[] =
 "  -T, --trigger=FUNC@act[,act,...]\n"
 "                             Trigger action on those FUNCs\n"
-"  -U, --unpatch=FUNC         Don't apply dynamic patching for FUNCs\n"
+"  -U, --unpatch=FUNC         Hide FUNCs from trace or don't apply dynamic patching for FUNCs\n"
```
I'd like to put it after the dynamic tracing as it's the primary feature.

namhyung commented on uftrace.c:
```diff
@@ -630,6 +630,7 @@ static int parse_option(struct opts *opts, int key, char *arg)
 
 	case 'U':
+		opts->unpatch = opt_add_string(opts->unpatch, arg);
 		opts->patch = opt_add_prefix_string(opts->patch, "!", arg);
```
This is redundant and I think we can use opts->patch (I know it's annoying).  You can ignore ones without the '!' prefix later.

namhyung commented on utils/filter.c:
```diff
@@ -640,6 +647,7 @@ static const struct trigger_action_parser actions[] = {
 	{ "depth=",    parse_depth_action,        TRIGGER_FL_FILTER, },
 	{ "time=",     parse_time_action,         TRIGGER_FL_FILTER, },
 	{ "caller",    parse_caller_action,       TRIGGER_FL_FILTER, },
+	{ "unpatch",   parse_unpatch_action,      TRIGGER_FL_FILTER, },
```
Can you please add "hide" too?

namhyung commented on utils/filter.h:
```diff
@@ -35,6 +35,7 @@ enum trigger_flag {
 	TRIGGER_FL_AUTO_ARGS	= (1U << 14),
 	TRIGGER_FL_CALLER	= (1U << 15),
 	TRIGGER_FL_SIGNAL	= (1U << 16),
+	TRIGGER_FL_UNPATCH	= (1U << 17),
```
We can make it _HIDE.

namhyung commented on utils/fstack.c:
```diff
@@ -327,6 +338,17 @@ static int setup_fstack_filters(struct uftrace_data *handle, char *filter_str,
+	if (unpatch_str) {
+		setting->private = unpatch_str;
+		walk_sessions(sessions, setup_unpatch, setting);
+		walk_sessions(sessions, count_filters, &count);
```
I think you should add count_unpatch (or count_hide).

honggyukim commented:
@namhyung I've renamed `-f`/`--output-fields` to `-O`/`--output-fields`, then used `-f` and `-n` for filter and notrace for the single function without affecting subtree.

`-n` option works both for record and replay, but `-f` works only for replay time yet. I need to update it to make `-f` work for record time as well.

namhyung commented:
OK, first we should update PR title accordingly.  And we are now not reusing the -P/-U options, so I'm thinking about this again... (sorry).

1. we only need this for the 'hide' case.  The single depth filter option can be specified like `-F foo@depth=1` : this will show foo function only and not its child functions.

2. the consistent filter experience is good, but I'm not sure we'll use 'filter-single' and 'notrace-single' for record.  Since it's better to use -P and -U (well, they'll support -pg binaries eventually) : they have same semantics and perform better.

3. so I'd like to implement `--hide` option for replay and take the `-H` from the `--host` (as it's seldom used).

honggyukim commented:
@namhyung Thanks for your feedback. Can we have a call during weekends to talk about this?

namhyung commented:
Sure!

honggyukim commented:
Hi @namhyung, I've just updated as you requested. The changes are as follows.
- Renive `-H` option from `--host`
- Use `-H`/`--hide` option
- It only works for replay time

The only problem left is to handle `{}` during replay as follows.
```
$ uftrace replay -H getpid
# DURATION     TID     FUNCTION
   0.793 us [ 21720] | __monstartup();
   0.415 us [ 21720] | __cxa_atexit();
            [ 21720] | main() {
            [ 21720] |   a() {
            [ 21720] |     b() {
            [ 21720] |       c() {
   0.803 us [ 21720] |       } /* c */
   1.044 us [ 21720] |     } /* b */
   1.298 us [ 21720] |   } /* a */
   1.541 us [ 21720] | } /* main */
```
Please have a look. Thanks.

namhyung commented on uftrace.c:
```diff
@@ -143,7 +144,7 @@ __used static const char uftrace_help[] =
 "  -F, --filter=FUNC          Only trace those FUNCs\n"
 "      --graphviz             Dump recorded data in DOT format\n"
-"  -H, --host=HOST            Send trace data to HOST instead of write to file\n"
+"      --host=HOST            Send trace data to HOST instead of write to file\n"
```
We should remove it from the doc too.

namhyung commented on tests/t252_filter_H.py:
```diff
@@ -0,0 +1,24 @@
+
+    def setup(self):
+        self.subcmd = 'replay'
+        self.option = '-H c'
```
It should work for 'live' command too.  Then we can just set self.option in the setup function and remove the preprare function.

namhyung commented on utils/fstack.c:
```diff
@@ -327,6 +338,17 @@ static int setup_fstack_filters(struct uftrace_data *handle, char *filter_str,
+	if (hide_str) {
+		setting->info_str = hide_str;
+		walk_sessions(sessions, setup_hide, setting);
+		walk_sessions(sessions, count_filters, &count_hide);
```
I don't think we can use count_filters here.  It just counts number of any entries, we need to count number of entries having the hide trigger.

honggyukim replied:
Please let me know if the updated patch is not what you wanted.

namhyung commented on utils/fstack.c:
```diff
@@ -442,13 +464,15 @@ int fstack_setup_filters(struct opts *opts, struct uftrace_data *handle)
-			       opts->caller ?: "");
+			       opts->caller ?: "",
+			       (((opts->filter || opts->trigger) && opts->caller) && opts->hide) ? " or " : "",
+			       opts->hide ?: "");
```
IIRC you suggested a cleaner way.. no?

honggyukim commented:
Thanks for the review. I've updated it again based on your comments.

namhyung reviewed (commented):
Looks good.  For better display I think we should merge functions contain hidden children properly.  Something like below:

```diff
diff --git a/utils/fstack.c b/utils/fstack.c
index 47d91f4c..1a717f32 100644
--- a/utils/fstack.c
+++ b/utils/fstack.c
@@ -813,7 +813,7 @@ static int fstack_check_skip(struct uftrace_task_reader *task,
        if (tr.flags & (TRIGGER_FL_DEPTH | TRIGGER_FL_TRACE_ON))
                return 1;
 
-       if (tr.flags & TRIGGER_FL_TRACE_OFF || depth <= 0)
+       if (tr.flags & (TRIGGER_FL_TRACE_OFF | TRIGGER_FL_HIDE) || depth <= 0)
                return -1;
 
        return 0;
```

honggyukim commented:
@namhyung Thanks for the tip. The change is applied and the output looks good as follows.
```
$ uftrace -H getpid t-abc
# DURATION     TID     FUNCTION
  41.600 us [  8009] | __monstartup();
   4.300 us [  8009] | __cxa_atexit();
            [  8009] | main() {
            [  8009] |   a() {
            [  8009] |     b() {
   5.900 us [  8009] |       c();
   8.800 us [  8009] |     } /* b */
  11.800 us [  8009] |   } /* a */
  14.700 us [  8009] | } /* main */
```

namhyung reviewed (approved):
LGTM

namhyung commented:
@yezz123 though it's a bit late, thanks for your review. :)
